src/mcp/server.ts

```diff
@@ -96,10 +102,30 @@ export function createServer(backend: ServerBackend): Server {
     }
   });
 
-  if (backend.serverInitialized)
-    server.oninitialized = () => backend.serverInitialized!(server.getClientVersion());
-  if (backend.serverClosed)
-    server.onclose = () => backend.serverClosed!();
-
+  addServerListener(server, 'initialized', () => backend.serverInitialized?.(server.getClientVersion()));
+  addServerListener(server, 'close', () => backend.serverClosed?.());
   return server;
 }
+
+const startHeartbeat = (server: Server) => {
+  const beat = () => {
+    Promise.race([
+      server.ping(),
+      new Promise((_, reject) => setTimeout(() => reject(new Error('ping timeout')), 5000)),
```

Comment:
I'd be more generous with the timeout, 15s maybe.

Reply:
Let's land as is and see. 5 seconds for roundtrip is a lot for local setup, we don't think it is used on server for now.